### PR TITLE
next: Support Switch Multiple Engine Heights

### PR DESCRIPTION
**Pull request by zsalch**
Closes: #356 

Support to switching multiple engine heights. (Currently only support Leela Zero Engine)

How to config:
1. In the config.txt file, add a itme (engine-command-list) to add some new command lines:

    "engine-command": "./leelaz --gtp --lagbuffer 0 --weights %network-file --threads 2",
    "engine-command-list": [
      "./leelaz --gtp --lagbuffer 0 --weights leelaz_172_b1c --threads 2",
      "./leelaz --gtp --lagbuffer 0 --weights human_best --threads 2",
      "./leelaz --gtp --lagbuffer 0 --weights elf_1_62b --threads 2",
      "./leelaz --gtp --lagbuffer 0 --weights leelaz_157_d35 --threads 2"
    ],

2. Use the shortcut key Ctrl+number to switch the weight and display the weight name in the left down, the engine command line will be displayed in the Title:

    Ctrl+0: Default switch to "engine-command"
    Ctrl+(1-9): Switch to "engine-command-list"

3. If the current node is in Branch, it also supports keeping the current node when switching weights.

Demo:
image
image

### CODE REVIEW

File: src/main/java/featurecat/lizzie/rules/Board.java [hunk 1]
```diff
@@ -433,6 +435,79 @@ public boolean nextMove() {
...
+            if (BoardHistoryList.isMainTrunk(node)) {
+                goToMoveNumber(moveNumber);
+            } else {
+                // If in Branch, restore by the back routing
+                goToMoveNumberByBack(moveNumber);
+            }
+        }
+    }
+    // Go to move number by back routing when in branch
+    public boolean goToMoveNumberByBack(int moveNumber) {
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/Board.java]: Why is this a separate method?

File: src/main/java/featurecat/lizzie/rules/Board.java [hunk 2]
```diff
@@ -433,6 +435,79 @@ public boolean nextMove() {
...
+            Leelaz.WinrateStats stats = Lizzie.leelaz.getWinrateStats();
+            if (stats.totalPlayouts >= history.getData().playouts) {
+                history.getData().winrate = stats.maxWinrate;
+                history.getData().playouts = stats.totalPlayouts;
+            }
+            return nextVariation(fromBackChildren);
+        }
+    }
+    // Save the move number
+    public BoardHistoryNode saveMoveNumber() {
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/Board.java]: The returned value is never used.

File: src/main/java/featurecat/lizzie/rules/Board.java [hunk 3]
```diff
@@ -433,6 +435,79 @@ public boolean nextMove() {
         }
     }
 
+    /**
+     * Goes to the next coordinate, thread safe
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/Board.java]: The documentation doesn't explain the meaning of the returned boolean.

File: src/main/java/featurecat/lizzie/rules/Board.java [hunk 4]
```diff
@@ -433,6 +435,79 @@ public boolean nextMove() {
...
+            if (!BoardHistoryList.isMainTrunk(curNode)) {
+                // If in Branch, save the back routing
+                saveBackRouting(curNode);
+            }
+            goToMoveNumber(0);
+        }
+        saveNode = curNode;
+        return curNode;
+    }
+    // Save the back routing
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/Board.java]: In general, I would avoid comments that paraphrase method name: they don't add any information to what the code already says.

File: src/main/java/featurecat/lizzie/rules/Board.java [hunk 5]
```diff
@@ -433,6 +435,79 @@ public boolean nextMove() {
...
+    public void saveBackRouting(BoardHistoryNode node) {
+        if (node != null && node.previous() != null) {
+            node.previous().setFromBackChildren(node.previous().getNexts().indexOf(node));
+            saveBackRouting(node.previous());
+        }
+        return;
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/rules/Board.java]: Useless return

File: src/main/java/featurecat/lizzie/rules/Board.java (hunk 1 above)
Comment (OlivierBlanvillain) [reply src/main/java/featurecat/lizzie/rules/Board.java]: Also the return value is never used, meaning the method can be refactored as follows:

```java
int delta = moveNumber - history.getMoveNumber();
for (int i = 0; i < Math.abs(delta); i++) {
    BoardHistoryNode curNode = history.getCurrentHistoryNode();
    if (curNode.numberOfChildren() > 1 && delta > 0) {
        nextMove(curNode.getFromBackChildren());
    } else {
        if (!(delta > 0 ? nextMove() : previousMove())) {
            return;
        }
    }
}
```

**Comment by zsalch** [comment]
@OlivierBlanvillain 
Thanks a lot for your detailed review, I will check and reply one by one.

File: src/main/java/featurecat/lizzie/Util.java [hunk 6]
```diff
@@ -80,4 +81,36 @@ public static void saveAsFile(URL url, File file) {
...
+     * @param fm
+     * @param fitWidth
+     * @return fitted
+     */
+    public static String fitStringByWidth(String line, FontMetrics fm, int fitWidth) {
+        if (line == null || line.length() == 0) {
+            return "";
+        }
+        int width = fm.stringWidth(line);
+        if (width  > fitWidth) {
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/Util.java]: Removed.

File: src/main/java/featurecat/lizzie/Util.java [hunk 7]
```diff
@@ -80,4 +81,36 @@ public static void saveAsFile(URL url, File file) {
             e.printStackTrace();
         }
     }
+
+    /**
+     * Fit the too long text by width
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/Util.java]: Changed.

File: src/main/java/featurecat/lizzie/analysis/Leelaz.java [hunk 8]
```diff
@@ -78,7 +90,8 @@ public Leelaz() throws IOException, JSONException {
         currentCmdNum = 0;
         cmdQueue = new ArrayDeque<>();
 
-        JSONObject config = Lizzie.config.config.getJSONObject("leelaz");
+        // Move config to member for other method call
+        config = Lizzie.config.config.getJSONObject("leelaz");
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/analysis/Leelaz.java]: for the function startEngine.

File: src/main/java/featurecat/lizzie/analysis/Leelaz.java [hunk 9]
```diff
@@ -87,6 +100,24 @@ public Leelaz() throws IOException, JSONException {
             updateToLatestNetwork();
         }
 
+
+        // command string for starting the engine
+        engineCommand = config.getString("engine-command");
+        // substitute in the weights file
+        engineCommand = engineCommand.replaceAll("%network-file", config.getString("network-file"));
+
+        // Init current engine no and start engine
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/analysis/Leelaz.java]: Changed.

File: src/main/java/featurecat/lizzie/analysis/Leelaz.java [hunk 10]
```diff
@@ -87,6 +100,24 @@ public Leelaz() throws IOException, JSONException {
...
+        currentEngineNo = 0;
+        startEngine(engineCommand);
+        Lizzie.frame.refreshBackground();
+    }
+
+    public void startEngine(String engineCommand) throws IOException {
+        // Check engine command
+        if (engineCommand == null || "".equals(engineCommand.trim())) {
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/analysis/Leelaz.java]: Changed.

File: src/main/java/featurecat/lizzie/analysis/Leelaz.java [hunk 11]
```diff
@@ -117,8 +162,43 @@ public Leelaz() throws IOException, JSONException {
...
-        Lizzie.frame.refreshBackground();
+        //new Thread(this::read).start();
+        //can stop engine for switching weights
+        executor = Executors.newSingleThreadScheduledExecutor();
+        executor.execute(this::read);
+    }
+
+
+    public void restartEngine(String engineCommand, int index) throws IOException {
+        if (engineCommand == null || "".equals(engineCommand.trim())) {
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/analysis/Leelaz.java]: Changed.

File: src/main/java/featurecat/lizzie/analysis/Leelaz.java [hunk 12]
```diff
@@ -117,8 +162,43 @@ public Leelaz() throws IOException, JSONException {
...
+            return;
+        }
+        isSwitching = " Switching";
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/analysis/Leelaz.java]: Changed.

File: src/main/java/featurecat/lizzie/gui/LizzieFrame.java [hunk 13]
```diff
@@ -881,8 +893,15 @@ public void toggleCoordinates() {
     }
 
     public void setPlayers(String whitePlayer, String blackPlayer) {
-        setTitle(String.format("%s (%s [W] vs %s [B])", DEFAULT_TITLE,
-                whitePlayer, blackPlayer));
+        // Display engine command in title
+        this.playerTitle = String.format("%s (%s [W] vs %s [B])", DEFAULT_TITLE,
+                whitePlayer, blackPlayer);
+        setTitle(this.playerTitle + (Lizzie.leelaz.engineCommand() != null ? " - " + Lizzie.leelaz.engineCommand() : ""));
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/gui/LizzieFrame.java]: I think can see the configuration of the command directly in the title.

File: src/main/java/featurecat/lizzie/rules/Board.java (hunk 1 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/Board.java]: The function goToMoveNumberByBack process the move by saved the children branch.

Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/Board.java]: Return value removed.

File: src/main/java/featurecat/lizzie/rules/Board.java (hunk 2 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/Board.java]: Removed.

File: src/main/java/featurecat/lizzie/rules/Board.java (hunk 3 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/Board.java]: Changed.

File: src/main/java/featurecat/lizzie/rules/Board.java (hunk 4 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/Board.java]: Changed.

File: src/main/java/featurecat/lizzie/rules/Board.java (hunk 5 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/Board.java]: Removed.

File: src/main/java/featurecat/lizzie/analysis/Leelaz.java [hunk 14]
```diff
@@ -95,13 +126,27 @@ public Leelaz() throws IOException, JSONException {
...
-        // command string for starting the engine
-        String engineCommand = config.getString("engine-command");
-        // substitute in the weights file
-        engineCommand = engineCommand.replaceAll("%network-file", config.getString("network-file"));
         // create this as a list which gets passed into the processbuilder
-        List<String> commands = Arrays.asList(engineCommand.split(" "));
+        commands = Arrays.asList(engineCommand.split(" "));
+
+        // get weight name
+        if (commands != null) {
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/analysis/Leelaz.java]: Changed part of logic to regular.

File: src/main/java/featurecat/lizzie/gui/LizzieFrame.java [hunk 15]
```diff
@@ -881,8 +893,15 @@ public void toggleCoordinates() {
...
+    }
+
+    public void setEngineTitle(String engineCommand) {
+        // Display engine command in title
+        setTitle((this.playerTitle != null ? this.playerTitle + " - " : "") + engineCommand);
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/gui/LizzieFrame.java]: Adjusted set title logic.

File: src/main/java/featurecat/lizzie/Util.java [hunk 16]
```diff
@@ -80,4 +81,37 @@ public static void saveAsFile(URL url, File file) {
             e.printStackTrace();
         }
     }
+
+    /**
```
Comment (OlivierBlanvillain) [inline src/main/java/featurecat/lizzie/Util.java]: Extra comment

**Comment by OlivierBlanvillain** [comment]
Needs a rebased, LGTM otherwise!

File: src/main/java/featurecat/lizzie/Util.java (hunk 16 above)
Comment (zsalch) [reply src/main/java/featurecat/lizzie/Util.java]: Removed.

**Comment by OlivierBlanvillain** [comment]
@zsalch They are still conflicts to be resolved, could you have a look?

**Comment by OlivierBlanvillain** [comment]
I took care of the conflicts, let's get this in :tada:

**Pull request merged**